kong/resty/mlcache/init.lua

```diff
@@ -888,6 +783,186 @@ function _M:get(key, opts, cb, ...)
 end
 
 
+function _M:renew(key, opts, cb, ...)
+    if not self.broadcast then
+        error("no ipc to propagate renew, specify opts.ipc_shm or opts.ipc", 2)
+    end
+
+    if type(key) ~= "string" then
+        error("key must be a string", 2)
+    end
+
+    -- opts validation
+    local ttl, neg_ttl, _, l1_serializer, shm_set_tries = check_opts(self, opts)
+
+    if type(cb) ~= "function" then
+        error("callback must be a function", 2)
+    end
+
+    -- restrict this key to the current namespace, so we isolate this
+    -- mlcache instance from potential other instances using the same
+    -- shm
+    local namespaced_key = self.name .. key
+
+    local v, shmerr = self.dict:get_stale(namespaced_key)
+    if v == nil then
+        if shmerr then
+            -- shmerr can be 'flags' upon successful get_stale() calls, so we
+            -- also check v == nil
+            return nil, "could not read from lua_shared_dict: " .. shmerr
+        end
+
+        -- if we specified shm_miss, it might be a negative hit cached
+        -- there
+        if self.dict_miss then
+            v, shmerr = self.dict_miss:get_stale(namespaced_key)
+            if v == nil and shmerr then
+                -- shmerr can be 'flags' upon successful get_stale() calls, so we
+                -- also check v == nil
+                return nil, "could not read from lua_shared_dict (miss): " .. shmerr
+            end
+        end
+    end
+
+    local version_before = get_version(shmerr or 0)
+
+    local lock, err = resty_lock:new(self.shm_locks, self.resty_lock_opts)
+    if not lock then
+        return nil, "could not create lock: " .. err
+    end
+
+    local elapsed, err = lock:lock(LOCK_KEY_PREFIX .. namespaced_key)
+    if not elapsed and err ~= "timeout"  then
+        return nil, "could not acquire callback lock: " .. err
+    end
+
+    local is_hit
+    local is_miss
+
+    v, shmerr = self.dict:get_stale(namespaced_key)
+    if v ~= nil then
+        is_hit = true
+
+    else
+        if shmerr then
+            -- shmerr can be 'flags' upon successful get_stale() calls, so we
+            -- also check v == nil
+            return nil, "could not read from lua_shared_dict: " .. shmerr
```

Comment:
hi @bungle do we need to unlock `lock` object here before it returns nil?

Reply:
Seems no bug here, from the lua-resty-lock doc:


> If the [unlock()](https://github.com/openresty/lua-resty-lock#unlock) method is never called after this method call, the lock will get released when
> 
> the current resty.lock object instance is collected automatically by the Lua GC.
the exptime for the lock entry is reached.



but I think we could `call unlock_and_ret()` to align with the `get()` api's action.